keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,539 @@
+
+- Intercept all Update operations against CRDs.
+- Prevent any changes to a CRD's `spec.versions[*].storage` fields if the CRD includes an annotation whose key is `ksvm.sigs.k8s.io/migrating`. 
+
+A summary of the proposed workflow can be found below:
+
+1. A `StorageVersionMigration` CR for the `foo.example.io` CRD has been created, possibly by the trigger-controller as [described here](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/acdee30ced218b79e39c6a701985e8cd8bd33824/USER_GUIDE.md#storage-version-migrator-in-a-nutshell)). 
+2. If a validatingWebhookConfiguration wih the `ksvm.sigs.k8s.io/storedVersionManagementEnabled: "true"` label exists steps 2-6 are enabled, otherwise fall back on existing behavior.
+3. The [kube-storage-version-migrator] adds the `ksvm.sigs.k8s.io/migrating=foo.example.io` annotation to the`foo.example.io` CRD.
+4. The [kube-storage-version-migrator] begins migrating all `foo` resources as it does today.
+5. When a migration is successful, the `foo` CRD's `status.storedVersion` is patched to only include the new stored version.
+6. When a migration is finished, the [kube-storage-version-migrator] removes the `ksvm.sigs.k8s.io/migrating=foo.example.io` annotation from the `foo.example.io` CRD.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I would like the [kube-storage-version-migrator] to automatically update a CRD's `status.storeVersions` after a successful a migration.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### CA Requirements
+
+The introduction of an admission webhook would require a valid CA.
+
+These certificates may be manually created, but it's possible that the CA will be managed by [cert-manager](https://github.com/jetstack/cert-manager) or other cert-provisioners. If the webhook's certificates are managed by means of a CRD (as is the case with cert-manager) a cyclical dependency is introduced in which:
+
+- The [kube-storage-version-migrator] manages the storage versions of CRD introduced by the cert provisioner.
+- The [kube-storage-version-migrator] relies on the cert provisioner for cert management.
+
+The above dependencies could cause an issue when:
+
+- A migration for one of the CRDs owned by the cert provisioner is started.
+- A new version of the cert provisioner is deployed that is not compatible with the CRD being migrated and introduces a new stored version.
+- The cert provisioner upgrade fails because of the admission webhook and is no longer able to provision new certificates.
+- The certificate for ksvm expires.
+
+In the situation described above, the cluster administrator would need to manually revert the cert provision upgrade to proceed.
+
+#### Alternative Approaches
```

Comment:
Like, if the CRD storage version didn't change during the migration, ksvm will update storedVersion; otherwise it waits for the next migration. Will that be sufficient?

Reply:
Thanks for the review @roycaihw!

> A fail-closed webhook that interprets all writes to CRDs (even when there is no ongoing migration) feel heavy.

In general, I agree that webhooks can be a bit heavy, but do CRDs experience a high volume of change requests? I was of the opinion that in general these resources are not changed on a regular cadence.

> I wonder if we've considered alternatives that don't involve webhooks. Like, if the CRD storage version didn't change during the migration, ksvm will update storedVersion; otherwise it waits for the next migration. Will that be sufficient?

I have two concerns here:
1. The "migration experience" is no longer deterministic. In instances where the stored version is not updated via the migration, the cluster admin must manual intervene or wait until a new stored version is introduced to the CRD. If a cluster admin needs to check if the stored version was updated, this proposal doesn't seem too valuable.
2. Short of a fail-closed webhook I don't know how we can guarantee that a resource will not be created at a version being removed. The alternative that comes to mind is watching the stored version at the throughout a migration, but there is still a small window of time between the point where the ksvm decides that the stored version did not change  and the time that the CRD stored versions is updated in which a new CR could be created, possibly resulting in data loss.

Reply:
If the storage version changed during a migration, by design ksvm will perform another migration. Unless the CRD keeps flip-flopping its storage version, storedVersion will be automatically updated eventually. I hope that addresses the UX concern. 

For the data loss concern, I agree we cannot resolve the race condition with what we have today. I imagine we need some API change to tell: what's the "storage version **generation**" before/after a migration. But not having to use a fail-closed webhook, which could be a single point of failure, seems a plus to me.

Reply:
>If the storage version changed during a migration, by design ksvm will perform another migration. Unless the CRD keeps flip-flopping its storage version, storedVersion will be automatically updated eventually. I hope that addresses the UX concern.

That's a great point, I hadn't considered that changing the stored version would immediately kick off a new migration.

>For the data loss concern, I agree we cannot resolve the race condition with what we have today. I imagine we need some API change to tell: what's the "storage version generation" before/after a migration. But not having to use a fail-closed webhook, which could be a single point of failure, seems a plus to me.

An interesting idea for an API, but even with this field a race condition existed within the gap between checking the `storage version generation` and the time that the storedVersion is removed from the CRD in which a CR at a removed stored version could be removed. I accept that this gap is a very small unit of time, but worth mentioning.

I would love to hear what @deads2k and @sttts have to say.